File: src/epay_tool/ai/gemini.py
"""Google Gemini implementation of the AI provider.

Uses structured output (response_schema) so every response is validated
JSON - no fragile text parsing. The free AI Studio tier is enough for
development; production use costs a few euros per month at 100+
products/day.
"""

from __future__ import annotations

import json
import mimetypes
from pathlib import Path

from google import genai
from google.genai import types

from epay_tool.ai import prompts
from epay_tool.ai.provider import AIProvider
from epay_tool.ai.schemas import ListingCopy, ListingReview, ProductIdentification
from epay_tool.config import Settings

MAX_IMAGE_BYTES = 6 * 1024 * 1024


class GeminiProvider(AIProvider):
    def __init__(self, settings: Settings):
        if not settings.gemini_api_key:
            raise RuntimeError("EPAY_GEMINI_API_KEY is not configured.")
        self.model = settings.gemini_model
        self.client = genai.Client(api_key=settings.gemini_api_key)

    # ---------------- interface ----------------

    def identify_product(self, photos, known=None, notes=None) -> ProductIdentification:
        parts: list = [self._image_part(p) for p in photos[:8]]
        parts.append(
            prompts.IDENTIFY_USER_TEMPLATE.format(
                known=json.dumps(known or {}, ensure_ascii=False),
                notes=notes or "-",
            )
        )
        return self._generate(parts, prompts.IDENTIFY_SYSTEM, ProductIdentification)

    def write_copy(self, facts) -> ListingCopy:
        content = "Produktdaten:\n" + json.dumps(facts, ensure_ascii=False, indent=2)
        return self._generate([content], prompts.COPY_SYSTEM, ListingCopy)

    def review_listing(self, draft) -> ListingReview:
        content = "Angebotsentwurf:\n" + json.dumps(draft, ensure_ascii=False, indent=2)
        return self._generate([content], prompts.REVIEW_SYSTEM, ListingReview)

    # ---------------- internals ----------------

    def _generate(self, contents: list, system: str, schema: type):
        response = self.client.models.generate_content(
            model=self.model,
            contents=contents,
            config=types.GenerateContentConfig(
                system_instruction=system,
                response_mime_type="application/json",
                response_schema=schema,
                temperature=0.2,
            ),
        )
        return schema.model_validate_json(response.text)

    @staticmethod
    def _image_part(path: Path) -> types.Part:
        data = path.read_bytes()
        if len(data) > MAX_IMAGE_BYTES:
            data = _downscale(path)
        mime = mimetypes.guess_type(str(path))[0] or "image/jpeg"
        return types.Part.from_bytes(data=data, mime_type=mime)


def _downscale(path: Path, max_side: int = 1600) -> bytes:
    """Shrink very large photos before upload - plenty for identification."""
    import io

    from PIL import Image

    img = Image.open(path)
    img.thumbnail((max_side, max_side))
    buf = io.BytesIO()
    img.convert("RGB").save(buf, format="JPEG", quality=85)
    return buf.getvalue()

File: src/main/ai/gemini.ts
/**
 * Google Gemini implementation of the AI provider.
 *
 * Uses JSON response mode and validates every reply with Zod, so a
 * malformed or surprising answer fails loudly here instead of quietly
 * corrupting a listing. The free AI Studio tier is enough for
 * development; production use costs a few euros per month at 100+
 * products/day.
 */

import { readFile } from 'node:fs/promises'
import { extname } from 'node:path'

import { GoogleGenAI, type Part } from '@google/genai'
import type { z } from 'zod'

import type { AppConfig } from '@main/config'

import { COPY_SYSTEM, IDENTIFY_SYSTEM, REVIEW_SYSTEM, identifyUserPrompt } from './prompts'
import type { AIProvider, IdentifyInput } from './provider'
import {
  listingCopySchema,
  listingReviewSchema,
  productIdentificationSchema,
  type ListingCopy,
  type ListingReview,
  type ProductIdentification,
} from './schemas'

const MAX_PHOTOS = 8

const MIME_TYPES: Record<string, string> = {
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.png': 'image/png',
  '.webp': 'image/webp',
}

export class GeminiProvider implements AIProvider {
  private readonly client: GoogleGenAI
  private readonly model: string

  constructor(config: AppConfig) {
    if (!config.geminiApiKey) {
      throw new Error('EPAY_GEMINI_API_KEY is not configured.')
    }
    this.client = new GoogleGenAI({ apiKey: config.geminiApiKey })
    this.model = config.geminiModel
  }

  async identifyProduct(input: IdentifyInput): Promise<ProductIdentification> {
    const imageParts: Part[] = await Promise.all(
      input.photoPaths.slice(0, MAX_PHOTOS).map(async (path) => ({
        inlineData: {
          data: (await readFile(path)).toString('base64'),
          mimeType: MIME_TYPES[extname(path).toLowerCase()] ?? 'image/jpeg',
        },
      })),
    )
    const parts: Part[] = [
      ...imageParts,
      { text: identifyUserPrompt(input.known ?? {}, input.notes ?? null) },
    ]
    return this.generate(parts, IDENTIFY_SYSTEM, productIdentificationSchema)
  }

  async writeCopy(facts: Record<string, unknown>): Promise<ListingCopy> {
    const parts: Part[] = [{ text: `Produktdaten:\n${JSON.stringify(facts, null, 2)}` }]
    return this.generate(parts, COPY_SYSTEM, listingCopySchema)
  }

  async reviewListing(draft: Record<string, unknown>): Promise<ListingReview> {
    const parts: Part[] = [{ text: `Angebotsentwurf:\n${JSON.stringify(draft, null, 2)}` }]
    return this.generate(parts, REVIEW_SYSTEM, listingReviewSchema)
  }

  private async generate<T extends z.ZodTypeAny>(
    parts: Part[],
    systemInstruction: string,
    schema: T,
  ): Promise<z.infer<T>> {
    const response = await this.client.models.generateContent({
      model: this.model,
      contents: [{ role: 'user', parts }],
      config: {
        systemInstruction,
        responseMimeType: 'application/json',
        temperature: 0.2,
      },
    })

    const text = response.text
    if (!text) throw new Error('The AI returned an empty response.')

    let parsed: unknown
    try {
      parsed = JSON.parse(text)
    } catch {
      throw new Error(`The AI returned invalid JSON: ${text.slice(0, 200)}`)
    }
    return schema.parse(parsed)
  }
}
